monkey/infection_monkey/network_scanning/mssql_fingerprinter.py

```diff
@@ -77,27 +80,51 @@ def _query_mssql_for_instance_data(host: str) -> Optional[bytes]:
         sock.close()
 
 
-def _get_services_from_server_data(data: bytes) -> Dict[str, Any]:
-    services = {MSSQL_SERVICE: {}}
-    services[MSSQL_SERVICE]["display_name"] = DISPLAY_NAME
-    services[MSSQL_SERVICE]["port"] = SQL_BROWSER_DEFAULT_PORT
+def _get_instance_info(instance_info: Sequence[str]) -> Dict[str, str]:
+    info = {}
+    it = iter(instance_info)
+    for k in it:
+        info[k] = next(it)
+
+    return info
+
+
+def _parse_instance(instance: str) -> Dict[str, str]:
+    instance_info = instance.split(";")
+    if len(instance_info) > 1:
+        return _get_instance_info(instance_info)
+
+    return {}
+
+
+def _get_services_from_server_data(data: bytes) -> List[DiscoveredService]:
+    services = []
+    discovered_network_service = NetworkService.UNKNOWN
 
     # Loop through the server data
+    # Example data:
+    # yServerName;MSSQL-16;InstanceName;MSSQLSERVER;IsClustered;No;Version;14.0.1000.169;tcp;1433;np;\\MSSQL-16\pipe\sql\query;;
     mssql_instances = filter(lambda i: i != "", data[3:].decode().split(";;"))
-
     for instance in mssql_instances:
-        instance_info = instance.split(";")
-        if len(instance_info) > 1:
-            services[MSSQL_SERVICE][instance_info[1]] = {}
-            for i in range(1, len(instance_info), 2):
-                # Each instance's info is nested under its own name, if there are multiple
-                # instances
-                # each will appear under its own name
-                services[MSSQL_SERVICE][instance_info[1]][instance_info[i - 1]] = instance_info[i]
-
-            logger.debug(f"Found MSSQL instance: {instance}")
-
-    if len(services[MSSQL_SERVICE].keys()) == 2:
-        services = {}
+        instance_info = _parse_instance(instance)
+        if "tcp" in instance_info:
+            # Here we are sure that UDP service is actualy MSSQL
+            discovered_network_service = NetworkService.MSSQL
+            services.append(
+                DiscoveredService(
+                    protocol=NetworkProtocol.TCP,
+                    port=NetworkPort(instance_info["tcp"]),
+                    services=discovered_network_service,
+                )
+            )
+        logger.debug(f"Found MSSQL instance: {instance}")
+
+    services.append(
+        DiscoveredService(
+            protocol=NetworkProtocol.UDP,
+            port=NetworkPort(SQL_BROWSER_DEFAULT_PORT),
+            services=discovered_network_service,
+        )
+    )
```

Comment:
We always add a UDP port? Even if tcp is in instance_info?

Reply:
Yep, Because we create UDP socket on line 45. Should we add only tcp ?

Reply:
In other words, we're only able to retrieve the `instance_info` after we successfully connect to a UDP port?

Reply:
Yep. Though I changed this a bit. Check the final commit

Reply:
The logic there actually looks backwards to me.

Whether or not to add the UDP port should depend on whether or not we got back a response that identifies a MSSQL server.